File: src/parser.py
from dataclasses import dataclass
from typing import List, Optional, Any, Dict
from .ast import (
    Program, Assign, Accumulate, TensorRef, FuncCall,
    AggregatorCall, Sum, Product, Number, Pragma
)
from .ast import Position


# --------------------------------------------------------------
# Token model (for illustration)
# --------------------------------------------------------------
@dataclass
class Token:
    type: str
    value: str
    line: int
    col: int


# --------------------------------------------------------------
# Parser core class
# --------------------------------------------------------------
class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0

    # ------------------------------
    # Helpers
    # ------------------------------
    def peek(self) -> Optional[Token]:
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def prev(self) -> Optional[Token]:
        return self.tokens[self.pos - 1] if self.pos > 0 else None

    def accept(self, typ: str, value: str = None) -> bool:
        t = self.peek()
        if t and t.type == typ and (value is None or t.value == value):
            self.pos += 1
            return True
        return False

    def expect(self, typ: str, value: str = None) -> Token:
        t = self.peek()
        if not t or t.type != typ or (value is not None and t.value != value):
            raise SyntaxError(f"Expected {typ}{'='+value if value else ''} at {t.line}:{t.col}")
        self.pos += 1
        return t

    def parse_identifier(self) -> str:
        t = self.expect("ID")
        return t.value

    # ------------------------------
    # Entry point
    # ------------------------------
    def parse_program(self) -> Program:
        stmts = []
        while True:
            tok = self.peek()
            if not tok or tok.type == "EOF":
                break
            if tok.type == "HASH":
                stmts.append(self.parse_pragma())
            else:
                stmts.append(self.parse_statement())
        return Program(stmts)

    # ------------------------------
    # Statement dispatch
    # ------------------------------
    def parse_statement(self):
        term = self.parse_term()

        if isinstance(term, (AggregatorCall, FuncCall)):
            return term

        if not isinstance(term, TensorRef):
            raise SyntaxError("Statement must begin with tensor reference, function, or aggregator call")

        if self.accept("ACCUM"):  # +=, max=, avg=
            op = self.prev().value
            rhs = self.parse_expr()
            return Accumulate(term, op, rhs)
        self.expect("OP", "=")
        rhs = self.parse_expr()
        return Assign(term, rhs)

    # ------------------------------
    # Expressions
    # ------------------------------
    def parse_expr(self):
        term = self.parse_product()
        while self.accept("OP", "+"):
            rhs = self.parse_product()
            term = Sum(term, rhs)
        return term

    def parse_product(self):
        factors = [self.parse_term()]
        while True:
            tok = self.peek()
            if not tok or tok.type in {"OP", "COMMA", "RPAREN", "RBRACE", "ACCUM", "EOF"}:
                break
            if tok.type in {"ID", "NUMBER", "LPAREN"}:
                factors.append(self.parse_term())
            else:
                break
        if len(factors) == 1:
            return factors[0]
        return Product(factors)

    # ------------------------------
    # Term (tensor, function, aggregator, number)
    # ------------------------------
    def parse_term(self):
        tok = self.peek()
        if not tok:
            raise SyntaxError("Unexpected EOF")

        # Number literal
        if tok.type == "NUMBER":
            self.pos += 1
            return Number(float(tok.value))

        # Identifier-based term
        if tok.type == "ID":
            name = self.parse_identifier()

            # Aggregator: sum[i](expr)
            if self.accept("LBRACKET"):  # "["
                indices = self.parse_index_list()
                self.expect("RBRACKET")  # "]"
                if self.accept("LPAREN"):  # "("
                    args = self.parse_arg_list()
                    self.expect("RPAREN")  # ")"
                    return AggregatorCall(name, indices, args)
                return TensorRef(name, indices)

            # Function call: f(x)
            elif self.accept("LPAREN"):
                args = self.parse_arg_list()
                self.expect("RPAREN")
                return FuncCall(name, args)

            # Plain tensor reference
            return TensorRef(name, [])

        # Parenthesized expression
        if self.accept("LPAREN"):
            expr = self.parse_expr()
            self.expect("RPAREN")
            return expr

        raise SyntaxError(f"Unexpected token {tok.type} '{tok.value}' at {tok.line}:{tok.col}")

    # ------------------------------
    # Index / argument lists
    # ------------------------------
    def parse_index_list(self):
        indices = [self.parse_identifier()]
        while self.accept("COMMA"):
            indices.append(self.parse_identifier())
        return indices

    def parse_arg_list(self):
        args = [self.parse_expr()]
        while self.accept("COMMA"):
            args.append(self.parse_expr())
        return args

    def parse_tensor_ref(self):
        name = self.parse_identifier()
        if self.accept("LBRACKET"):
            indices = self.parse_index_list()
            self.expect("RBRACKET")
            return TensorRef(name, indices)
        return TensorRef(name, [])

    # ------------------------------
    # Pragma (#pragma ethical {...})
    # ------------------------------
    def parse_pragma(self):
        self.expect("HASH")
        self.expect("PRAGMA")
        kind = self.parse_identifier()
        self.expect("LBRACE")
        spec = self.parse_pragma_fields()
        self.expect("RBRACE")
        return Pragma(kind, spec)

    def parse_pragma_fields(self) -> Dict[str, str]:
        fields = {}
        while True:
            key = self.parse_identifier()
            self.expect("COLON")
            val_tok = self.expect("STRING")
            fields[key] = val_tok.value.strip('"')
            if not self.accept("COMMA"):
                break
        return fields

File: compiler/tl_runtime_v0_7_cpp/src/parser/parser.cpp
#include "tl/parser/parser.hpp"

#include <stdexcept>
#include <map>

namespace tl::parser {

Parser::Parser(const std::string& source) : lexer_(source) {
    current_ = lexer_.next();
    if (current_.type == TokenType::EndOfFile) {
        end_ = true;
    }
}

std::unique_ptr<ast::Program> Parser::parse_program() {
    auto program = std::make_unique<ast::Program>();
    while (!end_) {
        auto stmt = parse_statement();
        if (stmt) {
            program->statements.push_back(std::move(stmt));
        }
        if (current_.type == TokenType::EndOfFile) {
            end_ = true;
        }
    }
    return program;
}

Token Parser::consume(TokenType type, const std::string& context) {
    if (current_.type != type) {
        throw std::runtime_error("Unexpected token in " + context + " at line " +
                                 std::to_string(current_.line));
    }
    Token t = current_;
    current_ = lexer_.next();
    return t;
}

bool Parser::match(TokenType type) {
    if (current_.type == type) {
        current_ = lexer_.next();
        return true;
    }
    return false;
}

bool Parser::check(TokenType type) {
    return current_.type == type;
}

Token Parser::advance() {
    Token t = current_;
    current_ = lexer_.next();
    if (current_.type == TokenType::EndOfFile) {
        end_ = true;
    }
    return t;
}

std::unique_ptr<ast::Statement> Parser::parse_statement() {
    switch (current_.type) {
        case TokenType::Comment:
            return parse_comment();
        case TokenType::Pragma:
            return parse_pragma();
        case TokenType::Identifier: {
            // Could be assign/accumulate or function-like statement.
            // Look ahead by parsing lhs then deciding.
            return parse_assign_or_accumulate();
        }
        case TokenType::EndOfFile:
            return nullptr;
        default:
            throw std::runtime_error("Unexpected start of statement at line " +
                                     std::to_string(current_.line));
    }
}

std::unique_ptr<ast::Statement> Parser::parse_comment() {
    auto text = current_.lexeme;
    advance();
    auto stmt = std::make_unique<ast::CommentStmt>();
    stmt->text = std::move(text);
    return stmt;
}

std::unique_ptr<ast::Statement> Parser::parse_pragma() {
    advance();  // consume #pragma

    std::string name;
    std::map<std::string, std::string> fields;

    if (current_.type == TokenType::Identifier) {
        name = advance().lexeme;
    }

    if (current_.type == TokenType::LBrace) {
        advance();  // '{'
        fields = parse_pragma_fields();
        consume(TokenType::RBrace, "pragma body end");
    }

    auto stmt = std::make_unique<ast::PragmaStmt>();
    stmt->name = std::move(name);
    stmt->fields = std::move(fields);
    return stmt;
}

std::map<std::string, std::string> Parser::parse_pragma_fields() {
    std::map<std::string, std::string> fields;
    if (check(TokenType::RBrace)) {
        return fields;
    }
    while (true) {
        Token key = consume(TokenType::Identifier, "pragma field key");
        consume(TokenType::Colon, "pragma field colon");
        Token value = consume(TokenType::StringLiteral, "pragma field value");
        fields[key.lexeme] = value.lexeme;
        if (match(TokenType::Comma)) {
            continue;
        }
        break;
    }
    return fields;
}

std::unique_ptr<ast::Statement> Parser::parse_assign_or_accumulate() {
    Token id = consume(TokenType::Identifier, "statement head");

    // Function-call statement without indices.
    if (current_.type == TokenType::LParen) {
        auto call_expr = std::make_unique<ast::FunctionCall>();
        call_expr->callee = id.lexeme;
        advance();  // '('
        call_expr->args = parse_arg_list();
        consume(TokenType::RParen, "function call ')'");

        auto stmt = std::make_unique<ast::FunctionCallStmt>();
        stmt->call = std::move(*call_expr);
        return stmt;
    }

    std::vector<ast::Index> indices;
    if (current_.type == TokenType::LBracket) {
        advance();
        indices = parse_index_list();
        consume(TokenType::RBracket, "index list ']'");
    }

    // Aggregator-call statement: id[indexes](args)
    if (!indices.empty() && current_.type == TokenType::LParen) {
        advance();  // '('
        auto args = parse_arg_list();
        consume(TokenType::RParen, "aggregator call ')'");
        auto stmt = std::make_unique<ast::AggregatorCallStmt>();
        stmt->call.callee = id.lexeme;
        stmt->call.indices = std::move(indices);
        stmt->call.args = std::move(args);
        return stmt;
    }

    // Tensor reference LHS for assign/accumulate
    ast::TensorRef lhs;
    lhs.name = id.lexeme;
    lhs.indices = std::move(indices);

    if (current_.type == TokenType::Assign ||
    current_.type == TokenType::PlusAssign) {

    ast::AssignKind kind = ast::AssignKind::Assign;
    if (current_.type == TokenType::PlusAssign) {
        kind = ast::AssignKind::PlusAssign;
    }

    advance();
    auto rhs = parse_expr();

    auto stmt = std::make_unique<ast::AssignStmt>();
    stmt->lhs = std::move(lhs);
    stmt->kind = kind;
    stmt->rhs = std::move(rhs);
    return stmt;
}

if (current_.type == TokenType::MaxAssign ||
    current_.type == TokenType::AvgAssign) {

    ast::AccumulateKind kind = ast::AccumulateKind::Add;
    if (current_.type == TokenType::MaxAssign) {
        kind = ast::AccumulateKind::Max;
    } else if (current_.type == TokenType::AvgAssign) {
        kind = ast::AccumulateKind::Avg;
    }

    advance();
    auto rhs = parse_expr();

    auto stmt = std::make_unique<ast::AccumulateStmt>();
    stmt->lhs = std::move(lhs);
    stmt->kind = kind;
    stmt->rhs = std::move(rhs);
    return stmt;
}

    throw std::runtime_error("Expected assignment, accumulate, or call after identifier at line " +
                             std::to_string(current_.line));
}

ast::TensorRef Parser::parse_tensor_ref() {
    Token name_tok = consume(TokenType::Identifier, "tensor ref");
    ast::TensorRef ref;
    ref.name = name_tok.lexeme;
    if (match(TokenType::LBracket)) {
        ref.indices = parse_index_list();
        consume(TokenType::RBracket, "tensor ref closing bracket");
    }
    return ref;
}

std::vector<ast::Index> Parser::parse_index_list() {
    std::vector<ast::Index> indices;
    indices.push_back(parse_index());
    while (match(TokenType::Comma)) {
        indices.push_back(parse_index());
    }
    return indices;
}

ast::Index Parser::parse_index() {
    if (match(TokenType::Star)) {
        Token id = consume(TokenType::Identifier, "star index identifier");
        ast::Index idx;
        idx.kind = ast::Index::Kind::StarId;
        idx.name = id.lexeme;
        return idx;
    }
    if (current_.type == TokenType::Identifier) {
        Token id = advance();
        ast::Index idx;
        idx.kind = ast::Index::Kind::Id;
        idx.name = id.lexeme;
        return idx;
    }
    if (current_.type == TokenType::Number) {
        Token num = advance();
        ast::Index idx;
        idx.kind = ast::Index::Kind::Number;
        idx.number = std::stod(num.lexeme);
        return idx;
    }
    throw std::runtime_error("Invalid index at line " + std::to_string(current_.line));
}

std::unique_ptr<ast::Expression> Parser::parse_expr() {
    // expr ::= sum_expr | product_expr | function_call | aggregator_call | number
    // sum_expr already nests product_expr, so treat expr as sum_expr entry point.
    return parse_sum_expr();
}

std::unique_ptr<ast::Expression> Parser::parse_sum_expr() {
    auto left = parse_product_expr();
    while (current_.type == TokenType::Plus) {
        advance();
        auto right = parse_product_expr();
        auto bin = std::make_unique<ast::BinaryExpr>();
        bin->op = ast::BinaryOpKind::Add;
        bin->left = std::move(left);
        bin->right = std::move(right);
        left = std::move(bin);
    }
    return left;
}

std::unique_ptr<ast::Expression> Parser::parse_product_expr() {
    auto left = parse_term();
    while (current_.type == TokenType::Identifier || current_.type == TokenType::Number ||
           current_.type == TokenType::LParen) {
        // Implicit multiplication (Einstein summation style) as consecutive terms.
        auto right = parse_term();
        auto bin = std::make_unique<ast::BinaryExpr>();
        bin->op = ast::BinaryOpKind::Multiply;
        bin->left = std::move(left);
        bin->right = std::move(right);
        left = std::move(bin);
    }
    return left;
}

std::unique_ptr<ast::Expression> Parser::parse_term() {
    if (current_.type == TokenType::Number) {
        Token num = advance();
        auto lit = std::make_unique<ast::NumberLiteral>();
        lit->value = std::stod(num.lexeme);
        return lit;
    }
    if (current_.type == TokenType::Identifier) {
        // Could be tensor_ref, function_call, or aggregator_call
        // Look ahead for '(' => function/aggregator, '[' + '(' => aggregator
        Token id = advance();
        if (current_.type == TokenType::LBracket) {
            // aggregator_call ::= id "[" index_list "]" "(" arg_list ")"
            // If '(' follows bracket block, aggregator_call; else tensor_ref.
            std::vector<ast::Index> indices;
            advance();  // consume '['
            indices = parse_index_list();
            consume(TokenType::RBracket, "aggregator/tensor closing bracket");
            if (current_.type == TokenType::LParen) {
                auto call = std::make_unique<ast::AggregatorCall>();
                call->callee = id.lexeme;
                call->indices = std::move(indices);
                advance();  // '('
                call->args = parse_arg_list();
                consume(TokenType::RParen, "aggregator call ')'");
                return call;
            }
            auto tref = std::make_unique<ast::TensorRef>();
            tref->name = id.lexeme;
            tref->indices = std::move(indices);
            return tref;
        }
        if (current_.type == TokenType::LParen) {
            auto call = std::make_unique<ast::FunctionCall>();
            call->callee = id.lexeme;
            advance();  // '('
            call->args = parse_arg_list();
            consume(TokenType::RParen, "function call ')'");
            return call;
        }
        auto ident = std::make_unique<ast::TensorRef>();
        ident->name = id.lexeme;
        return ident;
    }
    if (current_.type == TokenType::LParen) {
        advance();
        auto expr = parse_expr();
        consume(TokenType::RParen, "grouped expression");
        return expr;
    }
    throw std::runtime_error("Unexpected term at line " + std::to_string(current_.line));
}

std::unique_ptr<ast::Expression> Parser::parse_function_call() {
    Token id = consume(TokenType::Identifier, "function call");
    consume(TokenType::LParen, "function call '('");
    auto args = parse_arg_list();
    consume(TokenType::RParen, "function call ')'");

    auto call = std::make_unique<ast::FunctionCall>();
    call->callee = id.lexeme;
    call->args = std::move(args);
    return call;
}

std::unique_ptr<ast::Expression> Parser::parse_aggregator_call() {
    Token id = consume(TokenType::Identifier, "aggregator call");
    consume(TokenType::LBracket, "aggregator '['");
    auto indices = parse_index_list();
    consume(TokenType::RBracket, "aggregator ']'");
    consume(TokenType::LParen, "aggregator '('");
    auto args = parse_arg_list();
    consume(TokenType::RParen, "aggregator ')'");

    auto agg = std::make_unique<ast::AggregatorCall>();
    agg->callee = id.lexeme;
    agg->indices = std::move(indices);
    agg->args = std::move(args);
    return agg;
}

std::vector<std::unique_ptr<ast::Expression>> Parser::parse_arg_list() {
    std::vector<std::unique_ptr<ast::Expression>> args;
    if (current_.type == TokenType::RParen) {
        return args;
    }
    args.push_back(parse_expr());
    while (match(TokenType::Comma)) {
        args.push_back(parse_expr());
    }
    return args;
}

}  // namespace tl::parser
